packages/core/crates/core-protocol/src/msg/processor.rs

```diff
@@ -120,9 +122,260 @@ where
+                ChainPacketComponents::Final { .. } | ChainPacketComponents::Forwarded { .. } => {
+                    Err(PacketError::LogicError("Must contain an outgoing packet type".into()))
+                }
+                ChainPacketComponents::Outgoing {
+                    packet,
+                    ticket,
+                    next_hop,
+                    ack_challenge,
+                } => {
+                    self.db
+                        .write()
+                        .await
+                        .store_pending_acknowledgment(ack_challenge, PendingAcknowledgement::WaitingAsSender)
+                        .await?;
+
+                    let mut payload = Vec::with_capacity(ChainPacketComponents::SIZE);
+                    payload.extend_from_slice(packet.as_ref());
+                    payload.extend_from_slice(&ticket.to_bytes());
+
+                    Ok(TransportPacket::Outgoing {
+                        next_hop: next_hop.to_peerid(),
+                        ack_challenge: ack_challenge.clone(),
+                        data: payload.into_boxed_slice(),
+                    })
+                }
+            },
+            Err(e) => Err(e),
+        }
+    }
+
+    async fn from_incoming(
+        &self,
+        data: Box<[u8]>,
+        pkt_keypair: &OffchainKeypair,
+        sender: &PeerId,
+    ) -> Result<TransportPacket> {
+        let components = ChainPacketComponents::from_incoming(&data, pkt_keypair, sender)?;
+
+        match components {
+            ChainPacketComponents::Final {
+                packet_tag,
+                ack_key,
+                previous_hop,
+                plain_text,
+                ..
+            } => {
+                let ack = Acknowledgement::new(ack_key, pkt_keypair);
+
+                Ok(TransportPacket::Final {
+                    packet_tag,
+                    previous_hop: previous_hop.to_peerid(),
+                    plain_text,
+                    ack,
+                })
+            }
+            ChainPacketComponents::Forwarded {
+                packet,
+                ticket,
+                ack_challenge,
+                packet_tag,
+                ack_key,
+                previous_hop,
+                own_key,
+                next_hop,
+                next_challenge,
+                path_pos,
+            } => {
+                let domain_separator =
+                    self.db
+                        .read()
+                        .await
+                        .get_channels_domain_separator()
+                        .await?
+                        .ok_or_else(|| {
+                            debug!("Missing domain separator");
+                            MissingDomainSeparator
+                        })?;
+
+                let previous_peer = previous_hop.to_peerid();
+                let next_peer = next_hop.to_peerid();
+
+                // START: channel = get_channel_from_to(packet_key, packet_key)
```

Comment:
Note that Packet key and On-chain key resolving will be from an in-memory structure (LRU cache) and therefore won't be a DB query. 
So moving the chain key <-> packet key resolving more deeply in the DB trait might not be the correct way to go...

Reply:
I wanted to isolate chain operations, not only DB operations.